pageserver/src/relish_storage/storage_sync.rs

```diff
@@ -0,0 +1,1561 @@
+    let mut latest_timelines_for_tenants = HashMap::with_capacity(remote_timelines.len());
+    for ((remote_tenant_id, remote_timeline_id), remote_timeline_data) in remote_timelines {
+        let (latest_timeline_id, timeline_metadata) = latest_timelines_for_tenants
+            .entry(remote_tenant_id)
+            .or_insert_with(|| (remote_timeline_id, remote_timeline_data.metadata.clone()));
+        if latest_timeline_id != &remote_timeline_id
+            && timeline_metadata
+                .as_ref()
+                .map(|metadata| metadata.disk_consistent_lsn)
+                < remote_timeline_data.disk_consistent_lsn()
+        {
+            *latest_timeline_id = remote_timeline_id;
+            *timeline_metadata = remote_timeline_data.metadata.clone();
+        }
+    }
+
+    latest_timelines_for_tenants
+        .into_iter()
+        .map(|(&tenant_id, (&timeline_id, _))| (tenant_id, timeline_id))
+        .collect()
+}
+
+async fn fetch_existing_uploads<
+    P: std::fmt::Debug,
+    S: 'static + RelishStorage<RelishStoragePath = P>,
+>(
+    relish_storage: &S,
+) -> anyhow::Result<HashMap<(ZTenantId, ZTimelineId), RemoteTimeline>> {
+    let uploaded_relishes = relish_storage
+        .list_relishes()
+        .await
+        .context("Failed to list relish uploads")?;
+
+    let mut relish_data_fetches = uploaded_relishes
+        .into_iter()
+        .map(|remote_path| async {
+            (
+                remote_relish_data(relish_storage, &remote_path).await,
+                remote_path,
+            )
+        })
+        .collect::<FuturesUnordered<_>>();
+
+    let mut fetched = HashMap::new();
+    while let Some((fetch_result, remote_path)) = relish_data_fetches.next().await {
+        match fetch_result {
+            Ok((relish_info, remote_metadata)) => {
+                let tenant_id = relish_info.tenant_id;
+                let timeline_id = relish_info.timeline_id;
+                let remote_timeline =
+                    fetched
+                        .entry((tenant_id, timeline_id))
+                        .or_insert_with(|| RemoteTimeline {
+                            tenant_id,
+                            timeline_id,
+                            layers: Vec::new(),
+                            metadata: None,
+                        });
+                if remote_metadata.is_some() {
+                    remote_timeline.metadata = remote_metadata;
+                } else {
+                    remote_timeline
+                        .layers
+                        .push(relish_info.download_destination);
+                }
+            }
+            Err(e) => {
+                warn!(
+                    "Failed to fetch relish info for path {:?}, reason: {:#}",
+                    remote_path, e
+                );
+                continue;
+            }
+        }
+    }
+
+    Ok(fetched)
+}
+
+async fn remote_relish_data<P, S: 'static + RelishStorage<RelishStoragePath = P>>(
```

Comment:
should `metadata` be in the function name?

Reply:
I avoided using that, since we already have `TimelineMetadata` and adding another concept for layers seems a bit excessive.
Could use a different name, what about `remote_relish_info`, since that's the primary thing we return for relish here.